lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp

```diff
@@ -0,0 +1,261 @@
+///   vector<32xf32> vector_ext.yield %2 : vector<32xf32>
+/// }
+/// %0 = vector.transfer_read %src[%c0], %cst : memref<1024xf32>, vector<1xf32>
+struct WarpOpTransferRead : public OpRewritePattern<WarpSingleLaneOp> {
+  using OpRewritePattern<WarpSingleLaneOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(WarpSingleLaneOp warpOp,
+                                PatternRewriter &rewriter) const override {
+    auto operand = getWarpResult(
+        warpOp, [](Operation *op) { return isa<vector::TransferReadOp>(op); });
+    if (!operand)
+      return failure();
+    auto read = cast<vector::TransferReadOp>(operand->first);
+    Value distributedVal = warpOp.getResult(operand->second);
+
+    SmallVector<Value, 4> indices(read.indices().begin(), read.indices().end());
+    AffineMap map = calculateImplicitMap(read.getResult(), distributedVal);
+    AffineMap indexMap = map.compose(read.permutation_map());
+    OpBuilder::InsertionGuard g(rewriter);
+    rewriter.setInsertionPoint(warpOp);
+    for (auto it : llvm::zip(indexMap.getResults(), map.getResults())) {
+      AffineExpr d0, d1;
+      bindDims(read.getContext(), d0, d1);
+      auto indexExpr = std::get<0>(it).dyn_cast<AffineDimExpr>();
+      if (!indexExpr)
+        continue;
+      unsigned indexPos = indexExpr.getPosition();
+      unsigned vectorPos = std::get<1>(it).cast<AffineDimExpr>().getPosition();
+      auto scale = getAffineConstantExpr(
+          distributedVal.getType().cast<VectorType>().getDimSize(vectorPos),
+          read.getContext());
+      indices[indexPos] =
+          makeComposedAffineApply(rewriter, read.getLoc(), d0 + scale * d1,
+                                  {indices[indexPos], warpOp.laneid()});
+    }
+    Value newRead = rewriter.create<vector::TransferReadOp>(
+        read.getLoc(), distributedVal.getType(), read.source(), indices,
+        read.permutation_mapAttr(), read.padding(), read.mask(),
+        read.in_boundsAttr());
+    distributedVal.replaceAllUsesWith(newRead);
+    return success();
+  }
+};
+
+/// Remove any result that has no use along with the matching yieldOp operand.
+// TODO: Move this in WarpSingleLaneOp canonicalization.
+struct WarpOpDeadResult : public OpRewritePattern<WarpSingleLaneOp> {
+  using OpRewritePattern<WarpSingleLaneOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(WarpSingleLaneOp warpOp,
+                                PatternRewriter &rewriter) const override {
+    SmallVector<Type> resultTypes;
+    SmallVector<Value> yieldValue;
+    bool hasDeadResults = false;
+    auto yield = cast<vector_ext::YieldOp>(
+        warpOp.getBodyRegion().getBlocks().begin()->getTerminator());
+    for (OpResult result : warpOp.getResults()) {
+      if (result.use_empty()) {
+        hasDeadResults = true;
+        continue;
+      }
+      resultTypes.push_back(result.getType());
+      yieldValue.push_back(yield.getOperand(result.getResultNumber()));
+    }
+    if (!hasDeadResults)
```

Comment:
-> `if (yield.getNumOperands() == yieldValues.size())` and drop `hasDeadResults` everywhere